Remove BOOKMARKS_ENABLED feature flag

mebarbosa opened the pull request:
## Description
- Removes `BOOKMARKS_ENABLED` feature flag
- Removes early access logic for Bookmark feature
- Creates `BookmarkFeatureControl` to handle the feature's availability depending on the user's subscription
- Deletes `strings.xml` that we don't need anymore
- Clean up `WhatsNew` flow
- ⚠️ I need special attention for review: 4b17b2eb3aa53442e055300aa9d1faa50e21992b, b22960c8564a64fa7d8fd067ef07c06682bcd812 and e6b49f3eaa9752da3a77c5cd56274afbf97b9968 because I was not 100% what is the expected behavior for WhatsNew flow 

Closes #1688

## Testing In
git apply [debug.patch](https://github.com/Automattic/pocket-casts-android/files/13933869/debug.patchx) and run the app in `debug`

### 1 Testing with Patron or Plus user

#### 1.1 Headphone controls
-  Log in with a patron/plus user account
-  Go to Profile -> Settings -> Headphone controls -> Next Action
- [ ]  You should be able to add bookmark
-  Try with Previous action too
- [ ]  You should be able to add bookmark

#### 1.2 Podcast episode
- Play a podcast episode
- Click on the 3 dots of the menu
-  [ ]  You should be able to add bookmark
- In the same screen of the episode, click on the `Bookmarks` tab at the top
- [ ]  You should be able to see the bookmarks

#### 1.3 Podcast episode list
- Open an podcast episode list
- You will see the tabs at the top: `Episodes` and `Bookmarks`
- Click on Bookmarks
- [ ]  You should be able to see the bookmarks

### 2 Testing with free user

#### 2.1 Headphone controls
-  Log in with a patron/plus user account
-  Go to Profile -> Settings -> Headphone controls -> Next Action
- [ ]  You should not be able to add bookmark
- [ ]  The `Add bookmark` button must have an icon to indicates to upgrade to premium account
-  Try with Previous action too
- [ ]  You should not be able to add bookmark
- [ ]  The `Add bookmark` button must have an icon to indicates to upgrade to premium account

#### 2.2 Podcast episode
- Play a podcast episode
- Click on the 3 dots of the menu
-  [ ]  You should not be able to add bookmark
- In the same screen of the episode, click on the `Bookmarks` tab at the top
- [ ]  You should see the UpSell asking you to upgrade your account

#### 2.3 Podcast episode list
- Open an podcast episode list
- You will see the tabs at the top: `Episodes` and `Bookmarks`
- Click on Bookmarks
- [ ]  You should see the UpSell asking you to upgrade your account

#### 3 What's new modal

### 3.1 Free user existing account
- Uninstall the app 
- Increase the `Settings.WHATS_NEW_VERSION_CODE`
- Build and run the app again
- Log in with an existing free account
- [ ]  You should see the Bookmarks what's new modal with an button to upgrade your account like this:

<img src="https://github.com/Automattic/pocket-casts-android/assets/42220351/0a4d4dcf-eb35-47da-93eb-f86724bf38e2" height="400">

### 3.2 Patron/Plus user existing account
- Uninstall the app 
- Increase the `Settings.WHATS_NEW_VERSION_CODE`
- Build and run the app again
- Log in with an existing Patron/Plus account
- [ ]  You should see the Bookmarks what's new modal with an try button like this:

<img src="https://github.com/Automattic/pocket-casts-android/assets/42220351/5b3a7e7f-535b-476d-b4da-b635d3f826a0" height="400">

### 3.3 New account
- Uninstall the app 
- Increase the `Settings.WHATS_NEW_VERSION_CODE`
- Build and run the app again
- Create a new account
- [ ] You should not see the what's new modal


## Checklist
- [ ] If this is a user-facing change, I have added an entry in CHANGELOG.md
- [x] Ensure the linter passes (`./gradlew spotlessApply` to automatically apply formatting/linting)
- [x] I have considered whether it makes sense to add tests for my changes
- [x] All strings that need to be localized are in `modules/services/localization/src/main/res/values/strings.xml`
- [x] Any jetpack compose components I added or changed are covered by compose previews
 
#### I have tested any UI changes...
<!-- If this PR does not contain UI changes, ignore these items -->
- [ ] with different themes
- [ ] with a landscape orientation
- [ ] with the device set to have a large display and font size
- [ ] for accessibility with TalkBack


I would appreciate your inputs since you worked on this feature @ashiagr 🙏

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 500 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

mchowning commented on modules/features/player/src/main/java/au/com/shiftyjelly/pocketcasts/player/view/ShelfItem.kt:
```diff
@@ -22,9 +20,8 @@ object ShelfItems {
-        if (FeatureFlag.isEnabled(Feature.BOOKMARKS_ENABLED)) {
-            add(ShelfItem.Bookmark)
-        }
+        add(ShelfItem.Bookmark)
+
         add(ShelfItem.Archive)
```
```suggestion
        add(ShelfItem.Bookmark)
        add(ShelfItem.Archive)
```

mchowning reviewed (commented):
Really nice work on this @mebarbosa . Who would think that just removing a feature flag would be so difficult. This is looking really good though. I tested this and everything looks like it is behaving as it should.

I also did some testing of bookmarks with user files (Profile tab → Files) and confirmed they are still working there as expected too. 👍 

### 1. Testing What's New

When testing What's New I confirmed that the What's New dialog is not shown on a fresh install, and that it is shown with the appropriate button if I simulate an upgrade by bumping the `Settings.WHATS_NEW_VERSION_CODE`. This is a bit different from the test steps in the PR description, but I think it reflects the behavior in the production app (not showing what's new on fresh installs, but showing it on upgrades). So I think What's New is working as it should. 👍 

### 2. Removing early access logic

There are a few places where we're removing some early access logic because it's not being used anymore. I wonder if there might be a way we could restructure that logic a bit so that it remains, but is unused. My thinking is that we're going to need that logic again pretty soon, and it would be nice if we didn't have to try to recreate it. Adding new early access features and then removing their feature flags is something that I think we'll be doing quite a bit of, so it would be good if we could avoid adding and removing this logic for each feature.

On the other hand, it does feel weird to leave in unused logic, so let me know what you think.

mebarbosa commented on modules/features/player/src/main/java/au/com/shiftyjelly/pocketcasts/player/view/bookmark/components/UpsellViewModel.kt:
```diff
@@ -25,39 +20,20 @@ import kotlinx.coroutines.launch
     val state: StateFlow<UiState> = _state
 
     init {
-        val bookmarksFeature = feature.bookmarksFeature
-        val patronExclusiveAccessRelease = (bookmarksFeature.tier as? FeatureTier.Plus)?.patronExclusiveAccessRelease
-
-        val isReleaseCandidate = releaseVersion.currentReleaseVersion.releaseCandidate != null
-        val relativeToEarlyPatronAccess = patronExclusiveAccessRelease?.let {
-            releaseVersion.currentReleaseVersion.comparedToEarlyPatronAccess(it)
-        }
-        val availableForFeatureTier = when (relativeToEarlyPatronAccess) {
-            null -> bookmarksFeature.tier
-            EarlyAccessState.Before,
-            EarlyAccessState.During,
-            -> if (isReleaseCandidate) bookmarksFeature.tier else FeatureTier.Patron
-
-            EarlyAccessState.After -> bookmarksFeature.tier
-        }
-        val subscriptionTier = availableForFeatureTier.toSubscriptionTier()
-
         viewModelScope.launch {
-            subscriptionManager.freeTrialForSubscriptionTierFlow(subscriptionTier)
```
I had the same feeling about removing this. My initial thought was that since we use GitHub, we could simply copy and paste this logic from there as needed. However, I agree that this isn't the most convenient approach.

What just occurred to me is that we could extract this logic into a separate module. This way, we can easily remove or add it as needed. But I should create an issue for this since the current PR is already quite large. What do you think?

I can also keep this logic here to ensure we don't lose it

mebarbosa replied:
Ok, so I've created the `modules/services/utils/src/main/java/au/com/shiftyjelly/pocketcasts/utils/earlyaccess/EarlyAccessHelper.kt` to keep this logic and unit tests for it. Let me know what do you think

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 500 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

mebarbosa commented on modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/bookmark/BookmarkFeatureControl.kt:
```diff
@@ -0,0 +1,11 @@
+class BookmarkFeatureControl @Inject constructor() {
+    fun isAvailable(userTier: UserTier): Boolean =
+        userTier == UserTier.Patron || userTier == UserTier.Plus
+}
```
Yeah! I agree with you. The change is here: [5937a91](https://github.com/Automattic/pocket-casts-android/pull/1701/commits/5937a917cc3202126199b8b09008955855e9b68f)

mebarbosa commented on modules/services/utils/src/main/java/au/com/shiftyjelly/pocketcasts/utils/featureflag/Feature.kt:
```diff
@@ -47,22 +37,6 @@ enum class Feature(
     ),
     ;
 
-    fun isCurrentlyExclusiveToPatron(
-        releaseVersion: ReleaseVersionWrapper = ReleaseVersionWrapper(),
-    ): Boolean {
-        val isReleaseCandidate = releaseVersion.currentReleaseVersion.releaseCandidate != null
-        val relativeToEarlyAccessState = (this.tier as? FeatureTier.Plus)?.patronExclusiveAccessRelease?.let {
-            releaseVersion.currentReleaseVersion.comparedToEarlyPatronAccess(it)
-        }
-        return when (relativeToEarlyAccessState) {
-            null -> false
-            EarlyAccessState.Before,
-            EarlyAccessState.During,
-            -> !isReleaseCandidate
-            EarlyAccessState.After -> false
-        }
-    }
-
```
Yes, let's keep it. I added a comment to let everyone know that we can't delete it. [5dbce81](https://github.com/Automattic/pocket-casts-android/pull/1701/commits/5dbce81b399e688a7f31657e52434ecb3e136409)

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 500 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 500 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

mebarbosa commented on modules/features/profile/src/main/java/au/com/shiftyjelly/pocketcasts/profile/cloud/CloudFileBottomSheetFragment.kt:
```diff
@@ -224,7 +222,7 @@ class CloudFileBottomSheetFragment : BottomSheetDialogFragment() {
 
                 val layoutBookmark = binding.layoutBookmark
-                layoutBookmark.isVisible = FeatureFlag.isEnabled(Feature.BOOKMARKS_ENABLED)
+                layoutBookmark.isVisible = true
```
Nice catch! [f270f8d](https://github.com/Automattic/pocket-casts-android/pull/1701/commits/f270f8d3403e49db64af75b124ff146f636002da)

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 500 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

mebarbosa commented on modules/features/settings/src/main/java/au/com/shiftyjelly/pocketcasts/settings/HeadphoneControlsSettingsPageViewModel.kt:
```diff
@@ -134,22 +132,10 @@ class HeadphoneControlsSettingsPageViewModel @Inject constructor(
-            else -> null
-        }
+    private val addBookmarkIconColor: Color
+        get() = SubscriptionTierColor.plusGold
```
I prefer inline too! Done: [e656399](https://github.com/Automattic/pocket-casts-android/pull/1701/commits/e65639916e807224bae235f7f695b12593f23c3b)

mebarbosa commented on modules/services/preferences/src/main/java/au/com/shiftyjelly/pocketcasts/preferences/model/HeadphoneAction.kt:
```diff
@@ -32,9 +30,7 @@ class HeadphoneActionUserSetting(
-        val isAddBookmarkEnabled =
-            FeatureFlag.isEnabled(Feature.BOOKMARKS_ENABLED) &&
-                Feature.isUserEntitled(Feature.BOOKMARKS_ENABLED, userTier)
+        val isAddBookmarkEnabled = userTier == UserTier.Patron || userTier == UserTier.Plus
```
Done! [b348752](https://github.com/Automattic/pocket-casts-android/pull/1701/commits/b3487525fd5a9fb4ff01779e010dfd36dc69a6fa)

mebarbosa commented on modules/features/settings/src/main/java/au/com/shiftyjelly/pocketcasts/settings/whatsnew/WhatsNewViewModel.kt:
```diff
@@ -92,39 +72,14 @@ class WhatsNewViewModel @Inject constructor(
             subscriptionTier = subscriptionTier,
         )
     }
-
-    private fun titleResId(): Int {
-        val bookmarksFeature = feature.bookmarksFeature
-        val patronExclusiveAccessRelease = (bookmarksFeature.tier as? FeatureTier.Plus)?.patronExclusiveAccessRelease
-
-        val isReleaseCandidate = releaseVersion.currentReleaseVersion.releaseCandidate != null
-        val relativeToEarlyPatronAccess = patronExclusiveAccessRelease?.let {
-            releaseVersion.currentReleaseVersion.comparedToEarlyPatronAccess(it)
-        }
-
-        val showJoinBeta = when (relativeToEarlyPatronAccess) {
-            EarlyAccessState.Before,
-            EarlyAccessState.During,
-            -> isReleaseCandidate
-            EarlyAccessState.After -> false
-            null -> false
-        }
-
-        return if (showJoinBeta) {
-            LR.string.whats_new_boomarks_join_beta_testing_title
-        } else {
-            LR.string.whats_new_bookmarks_title
-        }
-    }
-
```
I extracted the logic as you suggested and added unit tests for this case. Looks better now 👌🏽 

See: [55b6278](https://github.com/Automattic/pocket-casts-android/pull/1701/commits/55b62788b9a9f3d10c8ff49d27c56ef1b45440eb)

mebarbosa commented:
hey @mchowning I addressed all your comments and suggestions. Thank you so much for a very detailed PR review. Would you mind to take a look at this again?

mchowning reviewed (approved):
Updates look great @mebarbosa ! 👍  Also retested and everything is still working well.
